### DISCUSSION TITLE
Export browser for Deno / Fresh & JSR

### DISCUSSION THREAD

**Discussion by hapaxlife** [MasterKale/SimpleWebAuthn#542] [Ideas]
Hi,

Thanks for your library :-)

You have export `server` & `type` for Deno. Could you do the same for `browser` in order to user it in Fresh ?

Do you plan to put it also on jsr ?

Thanks

**Comment by MasterKale** [comment]
Hello @hapaxlife, can you help refresh my memory what Fresh is?

As for JSR support, I'm intrigued but haven't sat down to figure out what it might take to migrate things from Deno + dnt for **server** and **types**. It also remains to be seen if JSR will see wide adoption.

I see that JSR is becoming [a first-class package repo in Deno 1.42](https://deno.com/blog/v1.42), so that's progress. And I know it's possible to [add JSR as a package repo for Node](https://jsr.io/docs/with/node). I'd say JSR is a strong contender for the next place I aim to start deploying SimpleWebAuthn to.

Yours is the first voice asking for deployments to alternative package repos than NPM and https://deno.land/x, but more people may chime in later. For now I'm content to wait and see how things play out.

**Comment by hapaxlife** [comment]
Thanks for your answer

Fresh is a full stack modern web framework for JavaScript and TypeScript developers. It is based on Deno and developped by Deno team.

It is a SSR with a concept of island for shipping javascript into page. For the client auth page, I import in my SSR Fresh the browser library (the origin of my initial demand). The server renders the page with the browser javascript code that is shipped to the client.
https://fresh.deno.dev/docs/concepts/islands

The fact that browser is not exported in Deno is not blocking. We have just different ways of import in the code.

**Comment by jpaoad31** [comment]
Hello @MasterKale,

I am chiming in because I am running into the same issue with writing client-side code in deno. I'm still very new to web development, so my apologies if anything I'm adding here is a little uninformed.

I am using Fresh in a web application and noticed there is no importable `browser.ts` available on `Deno.land/x/simplewebauthn`. Seems like an easy fix from my perspective, just include another "convenience import" file. If it is possible I would appreciate being able to use an import from Deno.land rather than having to rely on npm for this package.

Also while looking into this I noticed there are missing file extensions on imports throughout the entire `SimpleWebAuthn/packages/browser/src` directory. It prevents me from being able to import `SimpleWebAuthn/packages/browser/src/index.ts` directly into Deno without issues for missing import files. It seems like you added files extensions everywhere in the server source code in commit https://github.com/MasterKale/SimpleWebAuthn/commit/dc29a40014a3e750f1403ebab0b757d810468b6a so the lack of file extensions here appears to be a potential mistake or oversight.

**Reply by MasterKale** [reply]
I never targeted **@simplewebauthn/browser** for inclusion in 1) Deno projects that 2) leverage SSR. I can see how the structure of the browser package makes it difficult to use in such projects, though.

I'll have to think on this. I don't build the browser lib like I do server and types so I need to figure out how to write browser like Deno code (i.e. with `.ts` extensions in imports, etc... while still being able to build to single-file UMD and for NPM. I use [dnt](https://github.com/denoland/dnt) for building server and types, I wonder if it can work with Rollup 🤔

**Comment by MasterKale** [comment]
Hello @hapaxlife, @jpaoad31, @robvanvolt, and @sachaw, I'm pinging you all in here because you've all expressed varying degrees of interest in using SimpleWebAuthn in your Deno projects. I've been experimenting with hosting this project over on JSR.io in PR #634. I'm happy to report that I have "v11.0.0-alpha2" releases of **browser**, **server**, and **types** published over there:

![Screenshot 2024-11-08 at 10 50 16 PM](https://github.com/user-attachments/assets/0b8a57d4-114e-474a-8f47-ffffaf28ebd9)
https://jsr.io/@simplewebauthn

Can I ask for your help in putting these libraries through their paces? I'm fairly confident that `deno add jsr:@simplewebauthn/server` and `deno add jsr:@simplewebauthn/types` should work fine based on some of my own preliminary testing, but I'd love to hear about your experiences using these libraries as hosted through JSR.

`deno add jsr:@simplewebauthn/browser` should also be feasible, but I only published it just now and need to learn how to spin up a Fresh project to test it out with SSR myself. While I'm making time for that, those of you with active Fresh projects may be able to have some fun.

And heck, if you're so inclined, I'm also curious about your experiences using `npx jsr add` in any Node, Bun, and Cloudflare Worker projects you might be responsible for, or might spin up to test out.

JSR is tentatively a really cool way to host stuff, but I don't want to make it onerous for users to pull into their projects if I missed anything while experimenting with replacing the current deno.land/x hosting with JSR, if not also NPM hosting 🤔

**Reply by sachaw** [reply]
I'll have a play tonight if I get a chance.
May be worth looking at a few of the deno standardobrary packages that may be able to replace a few of the current dependencies. I have found them to be extremely well maintained.
https://jsr.io/@std/cbor
https://jsr.io/@std/encoding
As an example.

**Reply by robvanvolt** [reply]
Works flawless, I have just replaced every instance of npm: with the jsr: version - without deno add, I have just used the import x from y syntax and started the development server.

One thing I did not get running: the latest tag / automatic getting the latest version does not seem to work:

```
import {
  browserSupportsWebAuthn,
  startAuthentication,
} from "jsr:@simplewebauthn/browser@11.0.0-alpha2"; // specifying the version
```

does work.

```
import {
  browserSupportsWebAuthn,
  startAuthentication,
} from "npm:@simplewebauthn/browser"; // without specifying the version
```
npm does work as well even without specifying the exact version.


```
import {
  browserSupportsWebAuthn,
  startAuthentication,
} from "jsr:@simplewebauthn/browser
```

jsr does not work without specification:
error: Uncaught (in promise) TypeError: Could not find version of '@simplewebauthn/browser' that matches specified version constraint '*'

```
import {
  browserSupportsWebAuthn,
  startAuthentication,
} from "jsr:@simplewebauthn/browser@latest
```

jsr does also not work with latest tag or alike at the moment:
error: Uncaught (in promise) TypeError: Version tag not supported in jsr specifiers.

**Reply by MasterKale** [reply]
What does your import map entry for `@simplewebauthn/browser` look like after you ran `deno add jsr:@simplewebauthn/browser`? One thing I've observed about JSR is that it won't show anything on e.g. https://jsr.io/@simplewebauthn/browser because I have "not published a stable release yet":

![Screenshot 2024-11-09 at 9 31 27 AM](https://github.com/user-attachments/assets/246c396a-e492-470c-8260-0a0bd314a602)

I wonder if this is messing with package resolution, because JSR is saying there's no "latest" because there's no stable release yet 🤔

**Reply by MasterKale** [reply]
It's gotta be related to that, looking at the latest `deno add` commands in my bash history, I see an attempt to `deno add jsr:@simplewebauthn/server` then an immediate second attempt with specific version specifiers:

 
![Screenshot 2024-11-09 at 9 33 44 AM](https://github.com/user-attachments/assets/0d329d16-c5b1-4936-8fe8-e3aa9d1a8d07)

Funny enough I tried to `deno add jsr:@simplewebauthn/browser` just now and got an error that it wasn't found:

![Screenshot 2024-11-09 at 9 35 33 AM](https://github.com/user-attachments/assets/1c111a33-8e59-467d-844b-311e77efcaeb)

So I'll bet that's part of the pain here.

Interestingly, after a successful `deno add jsr:@simplewebauthn/browser@11.0.0-alpha2` (with the specific version identifier) my import map gets updated...

```json
{
  "tasks": {
    "dev": "deno run --watch main.ts"
  },
  "imports": {
    "@simplewebauthn/browser": "jsr:@simplewebauthn/browser@11.0.0-alpha2",
    "@simplewebauthn/server": "jsr:@simplewebauthn/server@11.0.0-alpha2",
    "@simplewebauthn/types": "jsr:@simplewebauthn/types@11.0.0-alpha2",
    "@std/assert": "jsr:@std/assert@1"
  }
}
```

...and the following works for me no problem:

**main.ts**

```ts
import { generateAuthenticationOptions } from '@simplewebauthn/server';
import { cose } from '@simplewebauthn/server/helpers'
import { AuthenticatorAttachment } from '@simplewebauthn/types';
import { base64URLStringToBuffer, browserSupportsWebAuthn } from '@simplewebauthn/browser';

/**
 * @simplewebauthn/types (typing works fine)
 */
const attachment: AuthenticatorAttachment = 'cross-platform';

/**
 * @simplewebauthn/server (method executes as expected)
 */
console.log(await generateAuthenticationOptions({
  rpID: 'example.com',
}));
console.log(cose.COSEKTY);

/**
 * @simplewebauthn/browser (all works as expected 🤯)
 */
console.log(base64URLStringToBuffer('AAAA'));
console.log(browserSupportsWebAuthn());
```

![Screenshot 2024-11-09 at 9 39 14 AM](https://github.com/user-attachments/assets/74234fca-6ea9-4221-8021-b304040c80c4)

**Reply by MasterKale** [reply]
Oh, wait, of course, `... from 'jsr:@simplewebauthn/browser'` is a totally valid way to import in Deno without using import maps 🤦‍♂️ 

I feel like that current issues with that import pattern and the alpha releases would get addressed if I simply published 11.0.0 as a stable release...

**Reply by MasterKale** [reply]
Something I'm also curious about: with Deno 2 adding support for `deno add npm:` do I "abandon" the idea of hosting on JSR and continue publishing ESM builds to NPM? What is the benefit of hosting to JSR if every runtime, Node, Deno, Bun, etc... supports installing packages from NPM?

**Reply by jpaoad31** [reply]
I switched my deno.land/x and npm imports to jsr and the authentication appears to be working fine. Although I am still very early into my project, so I will be sure to let you know if I run into any problems as I continue to work on it. I would hate to see you drop support for jsr already!

**Reply by MasterKale** [reply]
@jpaoad31 Can you help me understand your use case more? Is `deno add npm:` not feasible for your project (I'd guess maybe because you're not able to immediately switch to the latest and greatest Deno v2)? What is your perception of a project on JSR? Is it becoming something you prioritize when evaluating your third-party library options?

Hearing more from devs like you building things with SimpleWebAuthn helps me a lot to understand and plot out the future of this project. Any insights you can share with me about your use of Deno, of open source, and how working in Deno has changed your opinion about libraries would be much appreciated!

**Reply by sachaw** [reply]
One big thing for using JSR is uncompiled source distribution and automatic compilation for node/bun users.

**Reply by hapaxlife** [reply]
Hi,

Thank for trying to put your project on JSR. It works for my project (after migration to v11 😊 ). I have just switch from "npm" to "jsr" in my `deno.jsonc` file

```
   - "@simplewebauthn/browser": "npm:@simplewebauthn/browser@11.0.0",
   + "@simplewebauthn/browser": "jsr:@simplewebauthn/browser@11.0.0-alpha2",
```

**Reply by MasterKale** [reply]
Hey all, would it be acceptable to stop supporting import of these packages from deno.land/x and only support installation via `deno add jsr:@simplewebauthn/browser`/etc...? If I stop referencing third-party packages in SimpleWebAuthn via `https://...` specifiers (so I'm using normal-looking `import ... from 'somePackage'` within my libraries) because [JSR doesn't support `https://...` modules](https://jsr.io/docs/migrate-x-to-jsr#https-modules-supported-in-deno-not-in-jsr-packages), then importing SimpleWebAuthn like from `https://deno.land/x/simplewebauthn@v11.0.0-alpha3/deno/browser.ts` raises errors:

```
error: Relative import path "@hexagon/base64" not prefixed with / or ./ or ../ and not in import map from "https://deno.land/x/simplewebauthn@v11.0.0-alpha3/packages/server/src/helpers/iso/isoBase64URL.ts"
    at https://deno.land/x/simplewebauthn@v11.0.0-alpha3/packages/server/src/helpers/iso/isoBase64URL.ts:1:20
```

I can't see how I can support deno.land/x and JSR with the same codebase because of this fundamental restriction on packages that get hosted on JSR. `deno add jsr:@simplewebauthn/...` and `deno add npm:@simplewebauthn/...` are supported from Deno v1.46+ so perhaps it's not a big deal to say, "only installation from JSR and NPM in Deno projects is supported."

**Reply by MasterKale** [reply]
I'm going to drop deno.land/x hosting support. Deno's package hosting migration guidance is a little ambiguous but it definitely seems like it's "either JSR or deno.land/x" because of that restriction on using HTTPS module paths in JSR packages:

https://jsr.io/docs/migrate-x-to-jsr

I was going to release everything as "v12" anyway, but until now I didn't really have a reason to call it a major version release. "No longer support deno.land/x use" is a decent reason to increment the major version number.

**Reply by MasterKale** [reply]
cc @spendres

**Comment by MasterKale** [comment]
Progress report: I'm running some final tests on the big refactor I've been working on. Preliminary results are looking good, for example here's **@simplewebauthn/browser** running in a Deno Fresh project, installing "`v11.0.0-alpha3`" of the library from both JSR and deno.land/x:

![Screenshot 2024-11-24 at 2 41 04 PM](https://github.com/user-attachments/assets/ec0fc870-7cbe-4131-8a4f-aa0bc0d269cf)

**islands/WebAuthnJSR.tsx**
Run `deno add jsr:@simplewebauthn/browser@11.0.0-alpha3 jsr:@simplewebauthn/types@11.0.0-alpha3` then:

```tsx
import { browserSupportsWebAuthn, bufferToBase64URLString } from '@simplewebauthn/browser';
import { AuthenticatorAttachment } from '@simplewebauthn/types';

export default function WebAuthn() {
  const supportsWebAuthn = browserSupportsWebAuthn();
  const attachment: AuthenticatorAttachment = 'platform';

  return (
    <div class="py-6 items-center">
      <h1>@simplewebauthn/browser (JSR)</h1>
      <h2>WebAuthn supported?</h2>
      <h3>{`${supportsWebAuthn} (Client)`}</h3>
      <p>bufferToBase64URLString: {bufferToBase64URLString(new Uint8Array([0, 0, 0, 0]))}</p>
      <p>attachment: {attachment}</p>
    </div>
  );
}
```

~~**islands/WebAuthnDenoLand.tsx**~~
~~Specifying an `https://deno.land/x/...` path to the library:~~

```tsx
import { browserSupportsWebAuthn, bufferToBase64URLString } from "https://deno.land/x/simplewebauthn@v11.0.0-alpha3/deno/browser.ts";
import { AuthenticatorAttachment } from 'https://deno.land/x/simplewebauthn@v11.0.0-alpha3/deno/types.ts';

export default function WebAuthn() {
  const supportsWebAuthn = browserSupportsWebAuthn();
  const attachment: AuthenticatorAttachment = 'cross-platform';

  return (
    <div class="py-6 items-center">
      <h1>@simplewebauthn/browser (deno.land/x)</h1>
      <h2>WebAuthn supported?</h2>
      <h3>{`${supportsWebAuthn} (Client)`}</h3>
      <p>bufferToBase64URLString: {bufferToBase64URLString(new Uint8Array([0, 0, 0, 0]))}</p>
      <p>attachment: {attachment}</p>
    </div>
  );
}
```

EDIT: Importing from deno.land/x doesn't work anymore since I'm using import maps in @simplewebauthn/browser and @simplewebauthn/server...see https://github.com/MasterKale/SimpleWebAuthn/discussions/542#discussioncomment-11366189

Things are looking up...

**Comment by spendres** [comment]
I don't see a problem relying on jsr and dropping denoland. But, you could use deno.json import maps to transform your urls to names that vibe with jsr.

**Comment by spendres** [comment]
Oh, then you will have to abandon deno.land/x to support jsr.

**Comment by MasterKale** [comment]
Hello all, SimpleWebAuthn v12.0.0 marks the beginning of official support for installation of all three packages from JSR as well as from NPM 🚀 

See here for full release notes:

https://github.com/MasterKale/SimpleWebAuthn/releases/tag/v12.0.0

**Reply by hapaxlife** [reply]
Thanks you

**Comment by spendres** [comment]
Thank you, @MasterKale, for your unwavering effort to update this library.
